sled-agent/src/services.rs

```diff
@@ -525,117 +614,147 @@ impl ServiceManager {
+                    )?;
+                }
+                ServiceType::InternalDns { server_address, dns_address } => {
+                    info!(self.inner.log, "Setting up internal-dns service");
+                    smfh.setprop(
+                        "config/server_address",
+                        format!(
+                            "[{}]:{}",
+                            server_address.ip(),
+                            server_address.port(),
+                        ),
+                    )?;
+                    smfh.setprop(
+                        "config/dns_address",
+                        &format!(
+                            "[{}]:{}",
+                            dns_address.ip(),
+                            dns_address.port(),
+                        ),
+                    )?;
 
-                        smfh.setprop("config/pkt_source", pkt_source)?;
-                        smfh.refresh()?;
-                    }
+                    // Refresh the manifest with the new properties we set,
+                    // so they become "effective" properties when the service is enabled.
+                    smfh.refresh()?;
+                }
+                ServiceType::Oximeter => {
+                    info!(self.inner.log, "Setting up oximeter service");
+
+                    let address = request.addresses[0];
+                    smfh.setprop("config/id", request.id)?;
+                    smfh.setprop(
+                        "config/address",
+                        &format!("[{}]:{}", address, OXIMETER_PORT),
+                    )?;
+                    smfh.refresh()?;
+                }
+                ServiceType::Dendrite { asic } => {
+                    info!(self.inner.log, "Setting up dendrite service");
+
+                    let address = request.addresses[0];
+                    smfh.setprop("config/asic", asic)?;
+                    smfh.setprop(
+                        "config/address",
+                        &format!("[{}]:{}", address, DENDRITE_PORT,),
+                    )?;
+                    smfh.refresh()?;
                 }
+                ServiceType::Tfport { pkt_source } => {
+                    info!(self.inner.log, "Setting up tfport service");
 
-                debug!(self.log, "enabling service");
-                smfh.enable()?;
+                    smfh.setprop("config/pkt_source", pkt_source)?;
+                    smfh.refresh()?;
+                }
+            }
+
+            debug!(self.inner.log, "enabling service");
+            smfh.enable()?;
+        }
+        Ok(running_zone)
+    }
+
+    // Populates `existing_zones` according to the requests in `services`.
+    //
+    // At the point this function is invoked, IP addresses have already been
+    // allocated (by either RSS or Nexus). However, this function explicitly
+    // assigns such addresses to interfaces within zones.
+    async fn initialize_services_locked(
```

Comment:
This function is basically the same as it was, but the call to `self.initialize_zone` got refactored out.